### PR TITLE
Feature support non subcommand

### PR DESCRIPTION
**Pull request by Songmu**
By this patch, subcommands and main command can be coexistent. Empty string is considered as a main command.

### CODE REVIEW

**Comment by Songmu** [comment]
I want to use this package in https://github.com/mackerelio/mackerel-agent . For that, this feature should be merged.

**Comment by motemen** [comment]
Thanks. Mostly looks good, but I don't think the comment tag for the "main" action should not be `+command`, as it is not a "command" in the go-cli terminology. (Maybe `+main`, `+root` or some other tag?)

**Comment by Songmu** [comment]
I implemented `+main` tag. How about it?

BTW, `src`, third argument of `gen.Generate` seems not to be used for now. Do you have any idea apply it?

File: _example/main.go [hunk 1]
```diff
@@ -16,6 +16,16 @@ func main() {
 }
 
 /*
++main - generate example
+
```
Comment (motemen) [inline _example/main.go:20]: Sould have "usage" line like other subcommands like:

```
+main - ...

[tab]_example <command> [<args>...]

...
```

Comment (motemen) [inline _example/main.go:20]: Ah, you deliberately dropped it.Hm, seems ok, but you should document about the format in the `gen.Generate` doc.

File: cli.go [hunk 2]
```diff
@@ -129,7 +135,11 @@ func (app *App) PrintUsage() {
 
 	w := tabwriter.NewWriter(app.ErrorWriter, 0, 8, 4, ' ', 0)
 	for _, name := range names {
-		fmt.Fprintf(w, "    %s\t%s\n", name, app.Commands[name].Short)
+		n := name
+		if name == mainCmd {
+			n = "[main]"
```
Comment (motemen) [inline cli.go]: Wondering how we should describe this. "main" does not seem obvious to the users, so maybe `"<no command>"`?

File: cli.go [hunk 3]
```diff
@@ -73,29 +74,34 @@ func init() {
...
+// Dispatch is yet another entry point which returns error
+func (app *App) Dispatch(args []string) error {
+	cmdName, args := resolveCmd(args)
 	if cmd, ok := app.Commands[cmdName]; ok {
 		flags := flag.NewFlagSet(cmdName, app.FlagErrorHandling)
 		flags.Usage = func() {
 			fmt.Fprintln(app.ErrorWriter, cmd.Usage(flags))
+			if cmdName == mainCmd {
+				app.PrintUsage()
+			}
```
Comment (motemen) [inline cli.go:95]: Becomes odd if we add `flag.Parse(args)` inside `actionMain` in the example:

```
% ./_example -h
Usage: if no subcommands are specified, run it.
Usage: ./_example <command> [<args>]

Commands:
    [main]    generate example
    smile     show smile
    up        count up!
```

We should make `app.PrintUsage` go nicely with the main command.

There seem two options:
- Using documentation of the main command inside `app.PrintUsage`
- Change `flags.Usage` to call only `app.PrintUsage` if the command is the main command

File: cli_test.go [hunk 4]
```diff
@@ -124,7 +124,7 @@ blah
 	// prog cmd3
 	s = runOut("cmd3")
 	if strings.Contains(s, "internal error") == false {
-		t.Error("should include error inside action:\n%v", s)
+		t.Errorf("should include error inside action:\n%v", s)
```
Comment (motemen) [inline cli_test.go:127]: :+1:

**Comment by motemen** [comment]
Changes looks good, except for the usage issue :+1:

**Comment by motemen** [comment]
The argument `src` of `gen.Generate` was meant to be passed to `go/parser.ParseFile` so that codes can be generated from in-memory source, but it seemts that I failed to do so... :fearful:

**Comment by Songmu** [comment]
Fixed. Please review again.

**Comment by motemen** [comment]
Looks good, thanks :kissing:

**Pull request merged**